lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1172 @@
+		}
+		else{
+			return false;
+		}
+	}
+
+	// Useful method
+	bool ConstraintsLattice::foundInRanges(const std::pair<mlir::Value,bool>& key) const{
+		return this->ranges.find(key)!=this->ranges.end();
+	}
+
+	// Useful method
+	ConstraintsLattice::IntegerRange ConstraintsLattice::getTOP() const{
+		return createRange(MIN,MAX);
+	}
+
+	// Method for either inserting a value or joining the existing one 
+	bool ConstraintsLattice::insertOrJoin(const std::pair<mlir::Value,bool>& key, const ConstraintsLattice::IntegerRange& other){
+		auto found=this->ranges.find(key);
+		if(found!=this->ranges.end()){
+			auto copy=found->second;
+			found->second= 
+				ConstraintsLattice::joinRange(found->second, other);
+			return copy!=found->second;
+		}
+		else{
+			this->ranges.insert(std::make_pair(key, other));
+			return true;
+		}
+	}
+
+	// Method for either inserting a value or meeting the existing one 
+	bool ConstraintsLattice::insertOrMeet(const std::pair<mlir::Value,bool>& key, const ConstraintsLattice::IntegerRange& other){
+		auto found=this->ranges.find(key);
+		if(found!=this->ranges.end()){
+			auto copy=found->second;
+			found->second= 
+				ConstraintsLattice::meetRange(found->second, other);
+			return copy!=found->second;
+		}
+		else{
+			this->ranges.insert(std::make_pair(key, other));
+			return true;
+		}
+	}
+
+	// Method for either inserting a value or updating the existing one 
+	bool ConstraintsLattice::insertOrUpdate(const std::pair<mlir::Value,bool>& key, const ConstraintsLattice::IntegerRange& other){
+		auto found=this->ranges.find(key);
+		if(found!=this->ranges.end()){
+			auto copy=found->second;
+			found->second=other;
+			return copy!=found->second;
+		}
+		else{
+			this->ranges.insert(
+				std::make_pair(key, other));
+			return true;
+		}
+	}
+
+	// Method for either inserting a value or joining the existing one (overload)
+	bool ConstraintsLattice::insertOrUpdate(const std::pair<mlir::Value,bool>& keyToInsert, const std::optional<Integer>& min, const std::optional<Integer>& max){
+		auto found=this->ranges.find(keyToInsert);
+		if(found!=this->ranges.end()){
+			auto copy=found->second;
+			found->second=ConstraintsLattice::updateRange(found->second, min, max);
+			return copy!=found->second;
+		}
+		else{
+			this->ranges.insert(
+				std::make_pair(keyToInsert,
+					createRange(
+						min.value_or(MIN),
+						max.value_or(MAX))));
+			return true;
+		}
+	}
+
+	// Method for either inserting a value in the to_check structure
+	bool ConstraintsLattice::insertInTocheck(const mlir::Value& value){
+		return this->to_check.insert(value).second;
+	}
+
+	// This function is actually quite simpler since we always know 
+	// the lhs is not a constant and the rhs is a constant
+	// Result is useful to understand which check do to the operation
+	bool ConstraintsLattice::updateRelationalRange(mlir::Operation* statement, bool result, const mlir::Value& to_update, const mlir::Operation* other, const bool branch_to_update){
```

Comment:
can this be downgraded to a free function? ConstraintLattice stuff should not downcast operations to concrete types